### PR TITLE
[core][runtime_env] Always first do working_dir, and expose a env var for pip and conda

### LINKED ISSUE #18210: [runtime env] local pip "-r requirements.txt" not supported in conda spec
### What is the problem?
Currently the working directory is downloaded to the remote worker after the conda install has happened.  Thus, if the conda `environment.yml` contains a line under the `pip` field referencing a local file on the laptop, e.g. `-r laptop/requirements.txt` (which is supported by the conda spec), this file won't exist on the remote node at the time of the conda install and the installation will fail.

*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by rynewang**
## Why are these changes needed?

When you create a `pip` or `conda` runtime env, you can't reference the local requirements.txt files. For example this does not work:

environment.yml
```
dependencies:
  - pip
  - pip:
    - -r requirements.txt
```
run.py
```
ray.init(runtime_env:{"conda":"environment.yml", "working_dir":"."})
```

because
1. `working_dir` and `conda` are independent, there's no guarantee working_dir happens first
2. `conda` runs not in working_dir, but in the rt env agent. Similarly, `pip` runs in a purpose-created empty dir, not in the working_dir.

After some thoughts I decide to fix (1) and keep (2) because of implementation details (pip and conda copies the specs and writes to a specs file, and then calls the CLIs to install, hard to do that in working_dir). Instead, I introduce a new environment variable `RAY_RUNTIME_ENV_CREATE_WORKING_DIR` that is populated in the runtime envs' creation process.

Now, users have to be explicit on the referenced files, and it can work:

environment.yml
```
dependencies:
  - pip
  - pip:
    - -r ${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}/requirements.txt
```

Fixes #18210.

### CODE REVIEW

File: doc/source/ray-core/handling-dependencies.rst [hunk 1]
```diff
@@ -232,6 +232,8 @@ The specified local directory will automatically be pushed to the cluster nodes
 
 You can also specify files via a remote cloud storage URI; see :ref:`remote-uris` for details.
 
+If a `working_dir` is specified, it's always the first to be prepared, and it presents in the creation of other runtime envs in environment variable `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}`. This allows `pip` and `conda` to reference local files in the working_dir, e.g. `requirements.txt` or `environment.yml`.
```
Comment (angelinalg) [inline doc/source/ray-core/handling-dependencies.rst:235]:
```suggestion
If you specify a `working_dir`, Ray always prepares it first, and it's present in the creation of other runtime environments in the `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}` environment variable. This sequencing allows `pip` and `conda` to reference local files in the `working_dir` like `requirements.txt` or `environment.yml`.
```

File: doc/source/ray-core/handling-dependencies.rst [hunk 2]
```diff
@@ -389,7 +391,7 @@ The ``runtime_env`` is a Python dictionary or a Python class :class:`ray.runtime
   - Example: ``{"packages":["tensorflow", "requests"], "pip_check": False, "pip_version": "==22.0.2;python_version=='3.8.11'"}``
 
   When specifying a path to a ``requirements.txt`` file, the file must be present on your local machine and it must be a valid absolute path or relative filepath relative to your local current working directory, *not* relative to the `working_dir` specified in the `runtime_env`.
-  Furthermore, referencing local files `within` a `requirements.txt` file is not supported (e.g., ``-r ./my-laptop/more-requirements.txt``, ``./my-pkg.whl``).
+  Furthermore, referencing local files *within* a `requirements.txt` file is not directly supported (e.g., ``-r ./my-laptop/more-requirements.txt``, ``./my-pkg.whl``). Instead, a environment variable `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}` is provided in the creation process. You can use `-r ${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}/my-laptop/more-requirements.txt` or `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}/my-pkg.whl` to reference local files, as long as they're in the working_dir.
```
Comment (angelinalg) [inline doc/source/ray-core/handling-dependencies.rst:394]:
```suggestion
  Furthermore, referencing local files *within* a `requirements.txt` file isn't directly supported (e.g., ``-r ./my-laptop/more-requirements.txt``, ``./my-pkg.whl``). Instead, use the `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}` environment variable in the creation process. For example, use `-r ${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}/my-laptop/more-requirements.txt` or `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}/my-pkg.whl` to reference local files, while ensuring they're in the `working_dir`.
```

File: doc/source/ray-core/handling-dependencies.rst [hunk 3]
```diff
@@ -405,7 +407,7 @@ The ``runtime_env`` is a Python dictionary or a Python class :class:`ray.runtime
   - Example: ``"pytorch_p36"``
 
   When specifying a path to a ``environment.yml`` file, the file must be present on your local machine and it must be a valid absolute path or a relative filepath relative to your local current working directory, *not* relative to the `working_dir` specified in the `runtime_env`.
-  Furthermore, referencing local files `within` a `environment.yml` file is not supported.
+  Furthermore, referencing local files *within* a `environment.yml` file is not directly supported. Instead, a environment variable `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}` is provided in the creation process.
```
Comment (angelinalg) [inline doc/source/ray-core/handling-dependencies.rst:410]:
```suggestion
  Furthermore, referencing local files *within* a `environment.yml` file isn't directly supported. Instead, use the `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}`environment variable in the creation process.
```

File: python/ray/_private/ray_constants.py [hunk 4]
```diff
@@ -86,6 +86,9 @@ def env_set_by_user(key):
 RAY_RUNTIME_ENV_URI_PIN_EXPIRATION_S_ENV_VAR = (
     "RAY_RUNTIME_ENV_TEMPORARY_REFERENCE_EXPIRATION_S"
 )
+# This env var is populated to the working dir in the creation of a runtime env.
```
Comment (angelinalg) [inline python/ray/_private/ray_constants.py:89]:
```suggestion
# Ray populates this env var to the working dir in the creation of a runtime env.
```

File: python/ray/_private/ray_constants.py [hunk 5]
```diff
@@ -86,6 +86,9 @@ def env_set_by_user(key):
...
+# e.g. `pip` and `conda` users can use this to find their requirements.txt.
```
Comment (angelinalg) [inline python/ray/_private/ray_constants.py:90]:
```suggestion
# For example, `pip` and `conda` users can use this environment variable to locate the `requirements.txt` file.
```

**Review by angelinalg — APPROVED** [review]
Just some nits. Let me know if I've mangled the meaning and I can suggest a better rewrite. Generally, my suggestions are about removing the passive voice to clarify the subject of the verbs.

File: doc/source/ray-core/handling-dependencies.rst (hunk 3 above)
Comment (architkulkarni) [reply doc/source/ray-core/handling-dependencies.rst:410]: Nit: Let's also include your full description from above, it's pretty helpful.

> Instead, use the `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}` environment variable in the creation process. For example, use `-r ${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}/my-laptop/more-requirements.txt` or `${RAY_RUNTIME_ENV_CREATE_WORKING_DIR}/my-pkg.whl` to reference local files, while ensuring they're in the `working_dir`.

(Or whatever is required to make it work for `conda`.  It's just helpful to have a more fleshed out example.)

**Review by architkulkarni — APPROVED** [review]
Looks great!

**Comment by rynewang** [comment]
should work now. @architkulkarni please merge when you see this.

**Pull request merged**